### PR TITLE
core[minor]: add upsert, streaming_upsert, aupsert, astreaming_upsert methods to the VectorStore abstraction

### PR DESCRIPTION
**Pull request by eyurtsev**
This PR rolls out part of the new proposed interface for vectorstores (https://github.com/langchain-ai/langchain/pull/23544) to existing store implementations.

The PR makes the following changes:

1. Adds standard upsert, streaming_upsert, aupsert, astreaming_upsert methods to the vectorstore.
2. Updates `add_texts` and `aadd_texts` to be non required with a default implementation that delegates to `upsert` and `aupsert` if those have been implemented.  The original `add_texts`  and `aadd_texts` methods are problematic as they spread object specific information across document and **kwargs. (e.g., ids are not a part of the document)
3. Adds a default implementation to `add_documents` and `aadd_documents` that delegates to `upsert` and `aupsert` respectively.
4. Adds standard unit tests to verify that a given vectorstore implements a correct read/write API.

A downside of this implementation is that it creates `upsert` with a very similar signature to `add_documents`. 
The reason for introducing `upsert` is to:
* Remove any ambiguities about what information is allowed in `kwargs`. Specifically kwargs should only be used for information common to all indexed data. (e.g., indexing timeout). 
*Allow inheriting from an anticipated generalized interface for indexing that will allow indexing `BaseMedia` (i.e., allow making a vectorstore for images/audio etc.)
 
`add_documents` can be deprecated in the future in favor of `upsert` to make sure that users have a single correct way of indexing content.

### CODE REVIEW

File: libs/standard-tests/langchain_standard_tests/integration_tests/vectorstores.py [hunk 1]
```diff
@@ -250,25 +250,8 @@ async def test_add_documents_with_ids_is_idempotent(
         await vectorstore.aadd_documents(documents, ids=["1", "2"])
         documents = await vectorstore.asimilarity_search("bar", k=2)
         assert documents == [
-            Document(page_content="bar", metadata={"id": 2}),
-            Document(page_content="foo", metadata={"id": 1}),
-        ]
-
-    async def test_add_documents_without_ids_gets_duplicated(
```
Comment (eyurtsev) [inline libs/standard-tests/langchain_standard_tests/integration_tests/vectorstores.py:257]: Removed this unit test to allow vectorstores to implement assignment of IDs by document hash

File: libs/core/langchain_core/indexing/api.py [hunk 2]
```diff
@@ -405,7 +405,7 @@ async def _to_async_iterator(iterator: Iterable[T]) -> AsyncIterator[T]:
 
 
 async def aindex(
-    docs_source: Union[BaseLoader, Iterable[Document], AsyncIterator[Document]],
+    docs_source: Union[BaseLoader, Iterable[Document], AsyncIterable[Document]],
```
Comment (eyurtsev) [inline libs/core/langchain_core/indexing/api.py:408]:
```suggestion
    docs_source: Union[BaseLoader, Iterable[Document], AsyncIterator[Document]],
```

File: libs/core/langchain_core/indexing/base.py [hunk 3]
```diff
@@ -317,3 +317,16 @@ def delete_keys(self, keys: Sequence[str]) -> None:
...
+class UpsertResponse(TypedDict):
+    """A generic response for upsert operations.
+
+    The upsert response will be used by abstractions that implement an upsert
+    operation for content that can be upserted by ID.
+    """
+
+    succeeded: List[str]
+    """The IDs that were successfully indexed."""
+    failed: List[str]
```
Comment (ccurme) [inline libs/core/langchain_core/indexing/base.py:331]: (nit) are error messages useful to capture here

File: libs/core/langchain_core/vectorstores.py [hunk 4]
```diff
@@ -73,16 +85,205 @@ def add_texts(
...
+
+        .. versionadded:: 0.2.11
+        """
+        #  Developer guidelines:
+        #
+        #  Vectorstores implementations are free to extend `upsert` implementation
+        #  to take in additional data per document.
+        #
+        #  This data **SHOULD NOT** be part of the **kwargs** parameter, instead
+        #  a sub-classes can use a Union type on `documents` to include additional
```
Comment (ccurme) [inline libs/core/langchain_core/vectorstores.py:194]:
```suggestion
        #  sub-classes can use a Union type on `documents` to include additional
```

File: libs/core/langchain_core/vectorstores.py [hunk 5]
```diff
@@ -186,17 +387,81 @@ async def aadd_texts(
...
             List of IDs of the added texts.
         """
+        if type(self).upsert != VectorStore.upsert:
+            from langchain_core.documents import Document
+
+            if "ids" in kwargs:
+                ids = kwargs.pop("ids")
+                if ids and len(ids) != len(documents):
+                    raise ValueError(
+                        "The number of ids must match the number of documents."
```
Comment (ccurme) [inline libs/core/langchain_core/vectorstores.py:445]:
```suggestion
                        "The number of ids must match the number of documents. "
```

File: libs/core/langchain_core/vectorstores.py [hunk 6]
```diff
@@ -213,6 +478,38 @@ async def aadd_documents(
...
+                    doc_with_id = Document(
+                        page_content=document.page_content,
+                        metadata=document.metadata,
+                        id=id_,
+                    )
+                    documents_.append(doc_with_id)
+            else:
+                documents_ = documents
+
+            # If upsert has been implemented, we can use it to add documents
```
Comment (ccurme) [inline libs/core/langchain_core/vectorstores.py:509]: (nit) this comment is a little confusing as we only call `self.aupsert` (understood that `.aupsert` by default will use `.upsert`).

**Review by ccurme — APPROVED** [review]
just nits

File: libs/core/langchain_core/indexing/base.py (hunk 3 above)
Comment (eyurtsev) [reply libs/core/langchain_core/indexing/base.py:331]: No, just the IDs right now. We can expand later with a NotRequired key for error message? I'm not sure what useful just yet -- (an alternative is standardizing on exceptions)

File: libs/core/langchain_core/vectorstores.py (hunk 6 above)
Comment (eyurtsev) [reply libs/core/langchain_core/vectorstores.py:509]: Good point, I'll follow up with an update of the doc-string

**Pull request merged**